base/sparse/sparsevector.jl

```diff
@@ -421,12 +423,46 @@ function getindex{Tv}(A::SparseMatrixCSC{Tv}, I::AbstractVector)
     nzvalA = A.nzval
 
     n = length(I)
-    rowvalB = Array(Int, n)
-    nzvalB = Array(Tv, n)
+    nnzB = min(n, nnz(A))
+    rowvalB = Array(Int, nnzB)
+    nzvalB = Array(Tv, nnzB)
+
+    rowstart,colstart = ind2sub(szA, first(I))
+    rowend,colend = ind2sub(szA, last(I))
 
-    rowB = 1
     idxB = 1
+    @inbounds for col in colstart:colend
+        minrow = (col == colstart ? rowstart : 1)
+        maxrow = (col == colend ? rowend : szA[1])
+        for r in colptrA[col]:(colptrA[col+1]-1)
+            rowA = rowvalA[r]
+            if minrow <= rowA <= maxrow
+                rowvalB[idxB] = sub2ind(szA, rowA, col) - first(I) + 1
+                nzvalB[idxB] = nzvalA[r]
+                idxB += 1
+            end
+        end
+    end
+    if nnzB > (idxB-1)
+        deleteat!(nzvalB, idxB:nnzB)
+        deleteat!(rowvalB, idxB:nnzB)
+    end
+    SparseVector(n, rowvalB, nzvalB)
+end
+
+function getindex{Tv}(A::SparseMatrixCSC{Tv}, I::AbstractVector)
+    szA = size(A)
+    nA = szA[1]*szA[2]
+    colptrA = A.colptr
+    rowvalA = A.rowval
+    nzvalA = A.nzval
 
+    n = length(I)
+    nnzB = min(n, nnz(A))
```

Comment:
Need a better repository of real test code to benchmark sanely. I'd rather underallocate and grow an amortized number of times than overallocate for the case where many of the selected elements are zeros, but there will be corner cases for both.